### PR TITLE
Don't allow calling UnorderdMap::value_at for a set

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/4555.

### CODE REVIEW

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 1]
```diff
@@ -684,13 +695,28 @@ class UnorderedMap {
...
-  KOKKOS_FORCEINLINE_FUNCTION
-  std::conditional_t<(is_set || has_const_value), impl_value_type,
-                     impl_value_type &>
+  template <typename Dummy = value_type>
+  KOKKOS_FORCEINLINE_FUNCTION std::enable_if_t<
+      !std::is_void<Dummy>::value,  // !is_set
+      std::conditional_t<has_const_value, impl_value_type, impl_value_type &>>
   value_at(size_type i) const {
-    return m_values[is_set ? 0 : (i < capacity() ? i : capacity())];
+    KOKKOS_EXPECTS(i < capacity());
```
Comment (dalg24) [inline containers/src/Kokkos_UnorderedMap.hpp:703]: Same comment about change of behavior

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 2]
```diff
@@ -699,7 +725,8 @@ class UnorderedMap {
   /// kernel.
   KOKKOS_FORCEINLINE_FUNCTION
   key_type key_at(size_type i) const {
-    return m_keys[i < capacity() ? i : capacity()];
+    KOKKOS_EXPECTS(i < capacity());
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_UnorderedMap.hpp:728]: I would argue that returning the last element if out-of-range was just wrong/buggy. That was more like a drive-by change. If that is debatable I'm happy to just remove it.

[the remark this answers is not in the document]
Comment (dalg24) [reply containers/src/Kokkos_UnorderedMap.hpp:728]: https://github.com/kokkos/kokkos/blob/01974b7dd24388c1fc88e17c0f4788ed62811977/containers/src/Kokkos_UnorderedMap.hpp#L313

[the remark this answers is not in the document]
Comment (dalg24) [reply containers/src/Kokkos_UnorderedMap.hpp:728]: Not defending the current design, just pointing out that the behavior is changing

[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_UnorderedMap.hpp:736]: See b058e08f5.

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
@Char-Aznable would you please review?

**Review by dalg24 — APPROVED** [review]

**Comment by Char-Aznable** [comment]
Looks good to me! Thanks for the patch!

**Pull request merged**